bin/reth/src/chain/import.rs

```diff
@@ -0,0 +1,187 @@
+
+        init_genesis(db.clone(), self.chain.clone())?;
+
+        let consensus = self.init_consensus()?;
+        info!(target: "reth::cli", "Consensus engine initialized");
+
+        // create a new FileClient
+        let file_client = Arc::new(FileClient::new(&self.path).await?);
+
+        // construct downloaders and start pipeline
+        let mut pipeline = self.build_pipeline(&config, &file_client, &consensus, &db).await?;
+
+        tokio::spawn(handle_events(pipeline.events().map(Into::into)));
+
+        // Run pipeline
+        info!(target: "reth::cli", "Starting sync pipeline");
+        pipeline.run(db.clone()).await?;
+
+        info!(target: "reth::cli", "Finishing up");
+        Ok(())
+    }
+
+    async fn build_pipeline(
+        &self,
+        config: &Config,
+        file_client: &Arc<FileClient>,
+        consensus: &Arc<dyn Consensus>,
+        db: &Arc<Env<WriteMap>>,
+    ) -> eyre::Result<Pipeline<Env<WriteMap>, Arc<FileClient>>> {
+        let header_downloader =
+            self.spawn_headers_downloader(config, consensus, &file_client.clone());
+        let body_downloader =
+            self.spawn_bodies_downloader(config, consensus, &file_client.clone(), db);
+        let stage_conf = &config.stages;
+
+        let pipeline = Pipeline::builder()
+            .with_sync_state_updater(file_client.clone())
+            .add_stages(
+                OnlineStages::new(consensus.clone(), header_downloader, body_downloader).set(
+                    TotalDifficultyStage {
+                        commit_threshold: stage_conf.total_difficulty.commit_threshold,
+                    },
+                ),
+            )
+            .add_stages(
+                OfflineStages::default()
+                    .set(SenderRecoveryStage {
+                        batch_size: stage_conf.sender_recovery.batch_size,
+                        commit_threshold: stage_conf.execution.commit_threshold,
+                    })
+                    .set(ExecutionStage {
+                        chain_spec: self.chain.clone(),
+                        commit_threshold: stage_conf.execution.commit_threshold,
+                    }),
+            )
+            .build();
+
+        Ok(pipeline)
+    }
+
+    fn init_consensus(&self) -> eyre::Result<Arc<dyn Consensus>> {
+        let beacon_consensus = Arc::new(BeaconConsensus::new(self.chain.clone()));
+
+        // TODO: the node command requires a tip update, so we need to figure out how to do that
+        // with only a file
+        todo!()
+    }
+
+    fn spawn_headers_downloader(
+        &self,
+        config: &Config,
+        consensus: &Arc<dyn Consensus>,
+        file_client: &Arc<FileClient>,
+    ) -> reth_downloaders::headers::task::TaskDownloader {
+        // TODO: how to deal with the fact that this is a reverse downloader, and the blocks are
+        // written forwards in the file? RLP can only be parsed forwards
+        // do we need a forward downloader?
```

Comment:
Shouldn't matter - read all headers from the file in memory

Reply:
sounds good, will implement this